docs/docs/render-props.md

```diff
@@ -0,0 +1,224 @@
+
+As a first pass, you might try rendering the `<Cat>` *inside `<Mouse>`'s `render` method*, like this:
+
+```js
+class Cat extends React.Component {
+  render() {
+    const mouse = this.props.mouse
+    return (
+      <img src="/cat.jpg" style={{ position: 'absolute', left: mouse.x, top: mouse.y }}/>
+    );
+  }
+}
+
+class MouseWithCat extends React.Component {
+  constructor(props) {
+    super(props);
+    this.handleMouseMove = this.handleMouseMove.bind(this);
+    this.state = { x: 0, y: 0 };
+  }
+
+  handleMouseMove(event) {
+    this.setState({
+      x: event.clientX,
+      y: event.clientY
+    });
+  }
+
+  render() {
+    return (
+      <div style={{ height: '100%' }} onMouseMove={this.handleMouseMove}>
+
+        {/*
+          We could just swap out the <p> for a <Cat> here ... but then
+          we would need to create a separate <MouseWithSomethingElse>
+          component every time we need to use it, so <MouseWithCat>
+          isn't really reusable yet.
+        */}
+        <Cat mouse={this.state}/>
+      </div>
+    );
+  }
+}
+
+class MouseTracker extends React.Component {
+  render() {
+    return (
+      <div>
+        <h1>Move the mouse around!</h1>
+        <MouseWithCat/>
+      </div>
+    );
+  }
+}
+```
+
+This approach will work for our specific use case, but we haven't achieved the objective of truly encapsulating the behavior in a reusable way. Now, every time we want the mouse position for a different use case, we have to create a new component (i.e. essentially another `<MouseWithCat>`) that renders something specifically for that use case.
+
+Here's where the render prop comes in: Instead of hard-coding a `<Cat>` inside a `<Mouse>` component, and effectively changing its rendered output, we can provide `<Mouse>` with a function prop that it uses to dynamically determine what to render–a render prop.
+
+```js
+class Cat extends React.Component {
+  render() {
+    const mouse = this.props.mouse
+    return (
+      <img src="/cat.jpg" style={{ position: 'absolute', left: mouse.x, top: mouse.y }}/>
+    );
+  }
+}
+
+class Mouse extends React.Component {
+  constructor(props) {
+    super(props);
+    this.handleMouseMove = this.handleMouseMove.bind(this);
+    this.state = { x: 0, y: 0 };
+  }
+
+  handleMouseMove(event) {
+    this.setState({
+      x: event.clientX,
+      y: event.clientY
+    });
+  }
+
+  render() {
+    return (
+      <div style={{ height: '100%' }} onMouseMove={this.handleMouseMove}>
+
+        {/*
+          Instead of providing a static representation of what <Mouse> renders,
+          use the `render` prop to dynamically determine what to render.
+        */}
+        {this.props.render(this.state)}
+      </div>
+    );
+  }
+}
+
+class MouseTracker extends React.Component {
+  render() {
+    return (
+      <div>
+        <h1>Move the mouse around!</h1>
+        <Mouse render={mouse => (
```

Comment:
No need for a lambda here. Just inject the `Cat` component directly. Again, component injection.

`MouseTracker` should also be a SFC.
```js

const MouseTracker = () => (
  <div>
    <h1>Move the mouse around!</h1
    <Mouse render={Cat} />
  </div>
);
```

Reply:
There's nuance there, it's not being called as an element. People like to say components are "just functions" but how they get called matters. If it gets called as `React.createElement` you're gonna get a bunch of unexpected (or expected) mounting/unmounting. In this example its called as a function (not as createElement), so it would be semantically incorrect to pass the whole component in, that component is intended to be called with createElement.

Reply:
Rather than inlining a lambda, it might be worth making it a method since that's what you'd be more likely to see in a larger app:
```
<Mouse render={this._renderMouse} />
```

Reply:
Isn't this render function a good location for any prop mapping that may need to happen between `Mouse` and `Cat` (for example, converting `prop.mouse` into `prop.x` and `prop.y`? If we want to keep `Mouse` and `Cat` ignorant of one another, it seems odd to let `Mouse` directly render `Cat`. Also, what if you wanted to give `Cat` props in addition to those passed down by `Mouse`?

---

#### With render props:
```js
<Mouse render={mouse => (
  <Cat mouse={mouse} speed="slow" />
)}/>
```

#### With component injection:
```js
<Mouse render={Cat} /> // How do I set cat's speed? I wanted to keep Mouse alive. ;(
```

Reply:
p.s. I'm loving this example. It's so disarming. :D

Reply:
@donavon @Daniel15 
> Inline component definitions (like inline labmda function handlers in render) are just a bad idea from a memory management/garbage collection perspective.

If we don't have a benchmark demonstrating this, then it's a classic case of premature optimization. Inline definition is fine. If you measure your app and find bottlenecks, then optimize those components.

Reply:
@danny-andrews you could potentially pass through any unwanted props.  This is how `redux-form` does it.  However in this example, it just means that the mouse is eaten even quicker.

```jsx
<Mouse render={Cat} speed="slow" />
```